astro/setiathome/files/setiathome.1
.\" $FreeBSD: ports/astro/setiathome/files/setiathome.1,v 1.9 2000/04/17 01:08:33 kris Exp $
.Dd Dec 12, 2000
.Dt setiathome 1 LOCAL
.Os FreeBSD
.Sh NAME
.Nm setiathome
.Nd the SETI@home client program
.Sh SYNOPSIS
.Nm
.Op Ar options
.Sh DESCRIPTION
.Nm
is the
.Ux
version of the
.Tn SETI@home
client.
It downloads radio telescope data from a network server,
analyzes the data looking for signals of extraterrestrial origin,
and uploads results to the server, repeating this cycle indefinitely.
See
.Li http://setiathome.berkeley.edu
for more information.
.Pp
The first time you run
.Nm
it will interactively
ask you for E-mail address, name, country etc.
This info is stored in a file and no interaction is
needed when you run the program subsequently.
.Pp
After this you can run
.Nm
in the background,
and direct its output to
.Pa /dev/null
if you like.
.Pp
.Nm
can be freely stopped and restarted.
It saves its state in files, and will pick up where it left off.
.Pp
If you want to run multiple instances of
.Nm
(on a multiprocessor machine, or on multiple machines
that share a filesystem) each one must be run in a different directory.
.Sh IMPLEMENTATION NOTES
.Nm
requires about 0.5\ MBytes of disk space per working directory, and about
15\ MBytes of memory per instance.
If you have ample physical memory, its work load should be almost undetectable.
Additionally,
.Nm
requires a permanent or
.Dq dial on demand
Internet connection to upload results to and download new work units from the
.Tn SETI@home
server at any time.
.Sh OPTIONS
.Bl -tag -width countries
.It Fl countries
Show list of country codes.
.It Fl email
Send E-mail (to login E-mail address) on errors.
Useful if you run in background directed to
.Pa /dev/null .
.It Fl graphics
Generate a data stream for the
.Xr xsetiathome 1
graphical interface.
.It Fl login
Login or create new account.
.It Fl nice Ar N
Set
.Xr nice 1
priority to
.Ar N
(default 1).
.It Fl nolock
Omit the multiple-instance check, which uses file locking
(not available on some NFS systems).
.It Fl proxy Ar hostname : Ns Ar port
Connect to
.Tn SETI@home
server via specified HTTP proxy server and port.
.It Fl socks_passwd Ar password
SOCKS
.Ar password .
.It Fl socks_server Ar hostname : Ns Ar port
Connect to
.Tn SETI@home
server via specified SOCKS server and port.
Overrides
.Ev SOCKS_SERVER .
SOCKS versions 4 and 5 are supported.
.It Fl socks_user Ar name
SOCKS user
.Ar name .
.It Fl stop_after_process
Stop after processing current work unit.
Do not send result.
.It Fl stop_after_xfer
Stop after current work unit has been sent to the SETI
server and a new work unit has been obtained.
(See also
.Pa stop_after_send.txt
mentioned in
.Sx FILES
below.)
.It Fl verbose
Print a running summary of the work being done.
.It Fl version
Show software version.
.El
.Sh USAGE
.Ss Starting setiathome
This
.Fx
port includes a start-up script that starts
.Nm
at system start-up.
.Ss Configuration File
The optional configuration file
.Pa /usr/local/etc/setiathome.conf
can contain one or more of the following variable assignments to override
their default values:
.Bl -tag -width seti_wrkdir
.It Pa seti_wrkdir=/var/db/setiathome
The working directory where
.Nm
will store it's temporary files.
.It Pa seti_user=nobody
The user to run
.Nm
as.
.It Pa seti_nice=1
The
.Xr nice 1
value to run
.Nm 
at.
.It Pa seti_maxprocs= Ns Va Number of CPUs
The number of
.Nm 
processes to run concurrently. Defaults to the number of CPUs installed.
Running more than one instance per CPU is not useful, as
.Nm
is almost completely CPU-bound.
.El
.Pp
If you change values in the configuration file, you need to stop and start
.Nm Ns ,
and possibly reconfigure the working directory for the changes to take
effect. Run
.Pp
.Dl /usr/local/etc/rc.d/setiathome.sh stop
.Dl /usr/local/etc/rc.d/setiathome.sh register
.Dl /usr/local/etc/rc.d/setiathome.sh start
.Pp
to do that.
.Ss Setting up setiathome
You can use the start-up script to set up a working directory
and register with or log in to
.Tn SETI@home
by calling it with the argument
.Ar register .
If you'd like to configure and run it manually, you should first disable the
start-up script. Then, for each
.Nm
instance you'd like to run, create a directory, make it owned by the user
you'd like to run
.Nm
as, and run
.Ic setiathome -login
in that directory.
.Pp
Then, arrange for
.Nm
to be started automatically.  You can easily do so by adding a line similiar
to
.Bd -ragged -offset indent
.Li */10 * * * * cd 
.Va working directory 
.Li && setiathome -email >/dev/null
.Ed
.Pp
to the user's crontab.
.Sh ENVIRONMENT
.Bl -tag -width SOCKS_SERVER
.It Ev HTTP_PROXY
If defined,
.Nm
will connect through a proxy server, specified as
.Do Ns Va hostname Ns
.Op Li : Ns Va port
.Dc .
.It Ev SOCKS_SERVER
If defined,
.Nm
will connect through a SOCKS server, specified as
.Do Ns Va hostname Ns
.Op Li : Ns Va port
.Dc .
.It Ev SOCKS_USER , Ev SOCKS_PASSWD
If defined, these will be used as the login name and password
for the SOCKS server.
Otherwise
.Nm
will query you.
.El
.Sh FILES
.Bl -tag -width /var/db -compact
.It Pa /usr/local/etc/rc.d/setiathome.sh
Start-up script.
.Pp
.It Pa /usr/local/etc/setiathome.conf
Optional configuration file for the start-up script.
.Pp
.It Pa /var/db/setiathome/
Primary working directory.
.Pp
.It Pa /var/db/setiathome/user_info.sah
Registration data.
.Pp
.It Pa /var/db/setiathome/ Ns Va n Ns Pa /
Working directories for additional process
.Va n
(when running more than one instance).
.Pp
.It Pa /var/db/setiathome/ Ns Xo
.Op Ns Va n Ns Pa / Ns 
.Pa lock.sah
.Xc
Lock file to prevent multiple instances to work on the same data.
.Pp
.It Pa /var/db/setiathome/ Ns Xo
.Op Ns Va n Ns Pa / Ns 
.Pa *.sah
.Xc
Working files.
These should not be modified.
.Pp
.It Pa /var/db/setiathome/ Ns Xo
.Op Ns Va n Ns Pa / Ns 
.Pa pid.sah
.Xc
The process ID of the current instance.
.Pp
.It Pa /var/db/setiathome/ Ns Xo
.Op Ns Va n Ns Pa / Ns 
.Pa stop_after_send.txt
.Xc
While the client is running, touch this file in the working directory
to finish up your current work unit, return your result, and not
download a new work unit.
When the processing is finished, and the result sent, the client will stop.
Note this one file extension is
.Pa .txt ,
while all other files used by
the client have extensions
.Pa .sah .
.El
.Sh SEE ALSO
.Xr xsetiathome 1 ,
.Xr ipcs 1 ,
.Xr ipcrm 1 .
.Sh AUTHORS
.Tn SETI@home
was developed by David Anderson, Jeff Cobb, Charles Congdon, Charlie Fenton,
David Gedye, Kyle Granger, Eric Korpela, Matt Lebofsky, Peter Leiser, Brad
Silen, Woody Sullivan, and Dan Werthimer.
.Pp
.An Stefan Bethke Aq stb@freebsd.org
amended the original manual page for this
.Fx
port.
.Sh BUGS
Outstanding shared memory segments and semaphores may be left
active in case of an abnormal exit of the
.Ic setiathome -graphics
process.  These can prevent any future invocation of
.Ic setiathome -graphics .
To resolve this problem, use
.Xr ipcs 1
and
.Xr ipcrm 1
to remove shared memory segments and semaphores that
are not associated with a process.  This behavior may vary
depending upon how your
.Ux
system handles this situation.
.Pp
For version 3.* clients, the estimated progress as indicated
by the
.Pa prog=
line in the
.Pa state.sah
file in not exactly linear
in relationship to completion time.  Using this value to
predict completion time may not be completely accurate.
The linear relationship will vary depending upon the characteristics
of the work unit parameters.


CHANGES
Updating Information for FreeBSD ports developers

This file is maintained by Erwin Lansing <erwin@FreeBSD.org> and
copyrighted by the FreeBSD Foundation.

This file contains major changes to ports and the ports infrastructure.
Intended audience are ports committers, maintainers and other
developers. User oriented changes should be submitted for inclusion
in the release notes.

All ports committers are allowed to commit to this file.

20080905:
AUTHOR: hrs@FreeBSD.org

  * print/ghostscript-* and related ports have been renamed in the
    following way:
 
	print/ghostscript-gnu -> print/ghostscript7
	print/ghostscript-gnu-nox11 -> print/ghostscript7-nox11
	print/ghostscript-gnu-commfont -> print/ghostscript7-commfont
	print/ghostscript-gpl -> print/ghostscript8
	print/ghostscript-gpl-nox11 -> print/ghostscript8-nox11
	japanese/ghostscript-gnu-jpnfont -> print/ghostscript7-jpnfont
	korean/ghostscript-gnu-korfont -> print/ghostscript7-korfont

  * USE_GHOSTSCRIPT now supports a version number which the port
    requires.  The valid value is "7" or "8".  If other value is
    specified, value of WITH_GHOSTSCRIPT_VER is used.

  * WITH_GHOSTSCRIPT_GNU has been removed in favor of
    WITH_GHOSTSCRIPT_VER.  The valid value of WITH_GHOSTSCRIPT_VER is
    "7" or "8", and the default value is "8".

20080821:
AUTHOR: rafan@FreeBSD.org

   * CONFIGURE_ARGS is updated to use correct syntax for newer autoconf
     on the configure target. Individual ports no longer need to change
     CONFIGURE_TARGET to have the '--build=' prefix.

20080721:
AUTHOR: pav@FreeBSD.org

  * Default OpenLDAP version was changed from 2.3 to 2.4.

  * USE_GTK and USE_XPM variables are no longer recognized.

20080523:
AUTHOR: flz@FreeBSD.org

  * By default, ports-mgmt/pkg_install now sets up a wrapper to use ports
    pkg_install (when installed) instead of base when the former is more
    recent. This can be disabled in the options menu, or with the
    WITHOUT_WRAPPER knob.

20080414:
AUTHOR: pav@FreeBSD.org

  * @rmtry is now supported in pkg-plist.

  * USE_DISPLAY made more powerful, see comments in bsd.port.mk for details.

  * ${PW} is now defined.

  * LIB_DEPENDS can now be used with libraries with + sign in their name
    without the need for escaping.

20080312:
AUTHOR: pav@FreeBSD.org

  * USE_XPM no longer implies USE_XLIB.

  * USE_LDCONFIG can now be used to Linux binary ports that install shared
    libraries instead of INSTALLS_SHLIB.

20070908:
AUTHOR:	gabor@FreeBSD.org, linimon@FreeBSD.org

  * The Perl-related code parts have been extracted from bsd.port.mk to
    bsd.perl.mk.  This includes the user-settable knobs (e.g. USE_PERL5)
    and infrastructure parts of the code, like default targets.  The new
    code is designed to be conditionally included based on either
    USE_PERL5_*, PERL_CONFIGURE, or PERL_MODBUILD being set.  However,
    some of the default settings remain in bsd.port.mk while we make
    sure that all ports comply with this.

  * A convenient version handling has been implemented for Perl-related
    knobs.  With this new feature one can require a minimal, a maximal or an
    exact version, that the given port needs.  For the full description and
    some examples, please see the documentation in bsd.perl.mk.

20070804:
AUTHOR: gabor@FreeBSD.org
The following functional changes were made to bsd.*.mk and to
Tools/scripts/security-check.awk:

  * The variable definitions that cover command line tools, e.g. MKDIR, were
    extracted from bsd.port.mk into bsd.commands.mk, so that we can
    reuse them within the infrastructure later.

  * The old DESTDIR implementation was removed both from the infrastructure,
    and from the individual ports.  Variables like TARGETDIR should not longer
    be used.

  * A new DESTDIR implementation has been added, implemented in bsd.destdir.mk.
    The new implementation puts no requirements on the individual ports.

20070730:
AUTHOR: python@FreeBSD.org

  * The default Python version has been changed from 2.4.x to 2.5.x.

  * A number of variables have been added to bsd.python.mk to support
    the more integrated support for Python eggs, including:
    PYDISTUTILS_PKGNAME, PYDISTUTILS_PKGVERSION, PYEASYINSTALL_EGG,
    PYDISTUTILS_NOEGGINFO, PYDISTUTILS_EGGINFO, PYEASYINSTALL_ARCHDEP

  * More detailed descriptions of the variables can be found in the comments
    in bsd.python.mk.

20070723:
AUTHOR: rafan@FreeBSD.org

  * The default CONFIGURE_ARGS when GNU_CONFIGURE is set now contains
    --mandir and --infodir if configure script supports them.

20070712:
AUTHOR: maho@FreeBSD.org

  * Add knob USE_FORTRAN for ports which uses FORTRAN. All ports uses
    FORTRAN compiler should use this knob. Usages are: USE_FORTRAN=yes
    (default;gfortran42), USE_FORTRAN=ifc (Intel FORTRAN compiler)
    and USE_FORTRAN=g77 (/usr/bin/f77 or g77-34).

20070701:
AUTHOR: portmgr@FreeBSD.org
The following change was made to the ports infrastructure:

  * Virtual category 'kld' was added.  If the port installs kernel loadable
    modules, it should be included in this category.

20070524:
AUTHOR: portmgr@FreeBSD.org
The following changes were made to the ports infrastructure:

  * The 'make-deinstall-all' target now checks for moved ports.

  * The installation directories PORTEXAMPLES and PORTDATA are now defined.

  * The USE_MAKESELF knob is added for ports that use the makeself archiver.

  * The description of fetch-list was updated.  The targets fetch-required-list,
    fetch-url-list, and fetch-urlall-list were added.

  * 'make search' will also now search in ports/MOVED.

  * The default method for 'make update' is now portsnap.  Previously,
    you had to manually select one of 3 methods: SUP_UPDATE, CVS_UPDATE,
    or PORTSNAP_UPDATE.  The latter is now obsolete.

  * Several Makevar definitions have been moved to the pre-makefile section:
    DATADIR, DOCSDIR, ETCDIR, EXAMPLESDIR, WWWDIR.

20070403:
AUTHOR: portmgr@FreeBSD.org

astro/setiathome/files/setiathome.sh
#!/bin/sh
#
# $FreeBSD: ports/astro/setiathome/files/setiathome.sh,v 1.10 2003/04/09 05:02:42 cjc Exp $
#
# Start or stop SETI@home, or set up working directory and register.
#

case $0 in
/*) rc_dir=${0%/*} ;;
 *) rc_dir=${PWD:-$(pwd)} ;;
esac
rc_file=${0##*/}
rc_path=${rc_dir}/${rc_file}
rc_arg=$1

# override these variables in ${PREFIX}/etc/rc.setiathome.conf
seti_wrkdir=/var/db/${rc_file%.sh}	# primary working directory
seti_std_args=-email			# command arguments for standard mode
seti_reg_args=-login			# command arguments for register mode
seti_proxy_args=			# proxy arguments
seti_user=setiathome			# user id to run as
seti_group=${seti_user}			# group id to run as
seti_nice=15				# nice level to run at
seti_maxprocs=$(sysctl -n hw.ncpu)	# max. number of processes to start
seti_sleep=21600			# time to sleep between restarts

if ! PREFIX=$(expr ${rc_path} : "\(/.*\)/etc/rc\.d/${rc_file}\$"); then
	echo "${rc_file}: Cannot determine PREFIX." >&2
	echo "Please use the complete pathname." >&2
	exit 64
fi

rcconf_dir=${PREFIX}/etc
rcconf_file=rc.${rc_file%.sh}.conf
rcconf_path=${rcconf_dir}/${rcconf_file}

if [ -f ${rcconf_path} ]; then
	. ${rcconf_path}
fi

program_dir=${PREFIX}/sbin
program_file=${rc_file%.sh}
program_path=${program_dir}/${program_file}

syslog_facility=daemon.err

i=${seti_maxprocs}
seti_wrksuff="."
while [ ${i} -gt 1 ]; do
	seti_wrksuff="${seti_wrksuff} ${i}"
	i=$((${i} - 1))
done

case "$rc_arg" in
start)
	if [ ! -x ${program_path} ]; then
		logger -sp ${syslog_facility} -t ${program_file} \
			"unable to start: ${program_path} is missing."
		exit 72
	fi
	for i in ${seti_wrksuff}; do
		if [ ! -d ${seti_wrkdir}/${i} ]; then
			logger -sp ${syslog_facility} -t ${program_file} \
			"unable to start: ${seti_wrkdir}/${i} is missing."
			exit 72
		fi
		if [ ! -f ${seti_wrkdir}/${i}/user_info.sah ]; then
			logger -sp ${syslog_facility} -t ${program_file} \
			"unable to start: please log in to SETI@home first" \
			"(${seti_wrkdir}/${i}/user_info.sah is missing)."
			logger -sp ${syslog_facility} -t ${program_file} \
				"run \"$0 register\" to accomplish this."
			exit 72
		fi
	done
	for i in ${seti_wrksuff}; do
		su -fm ${seti_user} -c "exec /bin/sh -T" << EOF > /dev/null &
			cd ${seti_wrkdir}/${i} || exit
			echo \$\$ > shpid.sah
			trap 'kill \$pid;exit' 15
			while :; do
				${program_path} \
					${seti_std_args} ${seti_proxy_args} \
					${seti_nice:+-nice} ${seti_nice} &
				pid=\$!; wait \$pid
				sleep ${seti_sleep} &
				pid=\$!; wait \$pid
			done
EOF
	done
	echo -n " SETI@home"
	;;

stop)
	for i in ${seti_wrksuff}; do
		for pid_path in ${seti_wrkdir}/${i}/pid.sah \
				${seti_wrkdir}/${i}/shpid.sah; do
			if [ -f ${pid_path} ]; then
				kill $(cat ${pid_path}) 2> /dev/null
			fi
		done
	done
	if [ ! -f ${seti_wrkdir}/pid.sah ]; then
		killall ${program_file} 2> /dev/null
	fi
	;;

restart)
	$0 stop
	$0 start
	;;

status)
	ps auxww | egrep ${program_file} | egrep -v "($0|egrep)"
	;;

register)
	if [ ! -x ${program_path} ]; then
		echo "${program_file}:" \
			"unable to register: ${program_path} is missing." >&2
		exit 72
	fi
	if pw group show "${seti_group}" 2>/dev/null; then
		echo "You already have a group \"${seti_group}\"," \
		     "so I will use it."
	elif pw groupadd ${seti_group} -h -; then
		echo "Added group \"${seti_group}\"."
	else
		echo "Adding group \"${seti_group}\" failed..."
		echo "Please create it, and try again."
		exit 1
	fi
	if pw user show "${seti_user}" 2>/dev/null; then
		echo "You already have a user \"${seti_user}\"," \
		     "so I will use it."
		if pw usermod ${seti_user} -d ${seti_wrkdir}; then
			echo "Changed home directory of \"${seti_user}\"" \
			     "to \"${seti_wrkdir}\""
		else
			echo "Changing home directory of \"${seti_user}\"" \
			     "to \"${setu_wrkdir}\" failed..."
			exit 1
		fi
	elif pw useradd ${seti_user} -g ${seti_group} -h - \
		-d ${seti_wrkdir} -s /sbin/nologin -c "SETI at home Daemon"; then
		echo "Added user \"${seti_user}\"."
	else
		echo "Adding user \"${seti_user}\" failed..."
		echo "Please create it, and try again."
		exit 1
	fi
	# Create or update primary working directory (in case the uid changed)
	mkdir -p ${seti_wrkdir} || exit
	chown -Rh ${seti_user}:${seti_group} ${seti_wrkdir} || exit
	chmod u=Xrw,g=Xr,o=Xr ${seti_wrkdir} || exit
	seti_dontlogin=no
	if [ -f ${seti_wrkdir}/user_info.sah ]; then
		echo    "      It seems you have already registered with" \
			"SETI@home.  Would you"
		echo -n "      like to repeat the procedure? [Y/n] "
		read a
		if [ "X${a}" = "Xn" -o "X${a}" = "XN" ]; then
			seti_dontlogin=yes
		fi
	fi
	# No need to register if we've already done so
	if [ "X${seti_dontlogin}" != "Xyes" ]; then
		su -fm ${seti_user} -c "\
			cd ${seti_wrkdir} && \
			exec ${program_path} \
			${seti_reg_args} ${seti_proxy_args}"
	fi
	if [ ${seti_maxprocs} -gt 1 ]; then
		echo "      Updating additional working directories."
		i=${seti_maxprocs}
		while [ ${i} -gt 1 ]; do
			mkdir -p ${seti_wrkdir}/${i}
			chown ${seti_user} ${seti_wrkdir}/${i}
			chmod 755 ${seti_wrkdir}/${i}
			# Assume the user want's all processes to run
			# with the same registration
			rm -f ${seti_wrkdir}/${i}/user_info.sah
			ln -sf ../user_info.sah ${seti_wrkdir}/${i}
			i=$((${i} - 1))
		done
	fi
	;;		

*)
	echo "usage: ${rc_file} {start|stop|restart|status|register}" >&2
	exit 64
	;;
esac
